app/models/events/tags.rb

```diff
@@ -0,0 +1,40 @@
+module Events::Tags
+  extend ActiveSupport::Concern
+
+  def additional_type
+    tags.fetch(tag, nil)
+  end
+
+  def additional
+    additional_type&.new(super) || super
+  end
+
+  def additional=(value)
+    super(value&.to_h)
+  end
+
+  included do
+    validates :tag, presence: true, inclusion: { in: ->(_) { tags.keys } }
+    validates :additional, nested: true, allow_nil: true
+    belongs_to :ip_address, optional: true
+    belongs_to :geoip_info, optional: true
+
+    cattr_reader(:tags) { {} }
+  end
+
+  class_methods do
+    def define_event(tag, &blk)
```

Comment:
This all looks great. Thanks for hearing me out. You got it even simpler than I had imagined.